python/lsst/ts/watcher/rules/cp_verify_alarm.py

```diff
@@ -0,0 +1,296 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["CpVerifyAlarm"]
+
+import yaml
+import json
+import lsst.daf.butler as dafButler
+import salobj
+
+from lsst.ts.idl.enums.Watcher import AlarmSeverity
+from lsst.ts import watcher
+
+
+class CpVerifyAlarm(watcher.BaseRule):
+    """Monitor state of cp_verify executions from the OCPS
+    for a particular type of calibration (BIAS, DARK, FLAT, etc).
+
+    Set alarm severity NONE if the cp_verify tests passed,
+    and issue a WARNNING  alarm if not.
+
+    Parameters
+    ----------
+    config : `types.SimpleNamespace`
+        Rule configuration, as validated by the schema.
+
+    Notes
+    -----
+    The alarm name is f"CpVerifyAlarm.{ocps_index}:{calibration_type}",
+    where ``calibration_type`` and ``ocps_index`` are derived from
+    ``config.calibration_type`` and ``config.ocps_index``, respectively.
+    """
+
+    def __init__(self, config):
+        remote_name = "OCPS"
+        remote_index = config.ocps_index
+        remote_info = watcher.RemoteInfo(
+            name=remote_name,
+            index=remote_index,
+            callback_names=["job_result"],
+            poll_names=[],
+        )
+        super().__init__(
+            config=config,
+            name=f"CpVerifyAlarm.{remote_info.index}:{remote_info.name}",
+            remote_info_list=[remote_info],
+        )
+
+        self.calibration_type = config.calibration_type
+
+    @classmethod
+    def get_schema(cls):
+        schema_yaml = """
+            $schema: http://json-schema.org/draft-07/schema#
+            description: Configuration for CpVerifyBiasAlarm
+            type: object
+            properties:
+                calibration_type:
+                    description: >-
+                        Alarm calibration type: BIAS, DARK, or FLAT.
+                    type: string
+                ocps_index:
+                    OCPS index (e.g., OCPS:1: LATISS; OCPS:2: LSSTComCam).
+                verification_threshold:
+                    type: integer
+                    descriptor: Maximum number of failures per detector per test type.
+                default: 8
+            required: [calibration_type, ocps_index]
+            additionalProperties: false
+        """
+        return yaml.safe_load(schema_yaml)
+
+    def __call__(self, topic_callback):
+        msg = topic_callback.get()
+        # OCPS response after calling the pipetask
+        response = json.loads(msg.result)
+        # Get the dictionary with cp_verify stats, from the butler.
+        verify_stats = self.get_cp_verify_stats(response)
```

Comment:
This call may run slowly and block the Watcher from doing anything else. Please run it in a thread, e.g. using `run_in_executor`. This salobj documentation for writing scripts shows how: https://ts-salobj.lsst.io/sal_scripts.html#run-method

You can also create a concurrent.futures.ThreadPoolExecutor and pass that as the first argument to run_in_executor, instead of None. I suggest that, because you'll be calling this a lot and it is more efficient to create one and keep using it.

Which brings up a question: how often will this run? Will it have enough time to process one message before the next one comes in? In other words: how many butlers will be running queries at the same time? Just curious. if you create your own thread pool executor you can limit this (e.g. if you specify n=1 then only one will run at a time, which could make processing your rule back up, but will not block the Watcher itself, n=3 would allow up to 3 to run at the same time...).

Reply:
I have added `run_in_executor` and `concurrent.futures.ThreadPoolExecutor` when getting the verification statistics, thank you for the suggestions. 
 
The idea is to run one of the instances of "base_make_calibrations.py" via LOVE every afternoon to take individual calibrations, and use these individual images to run `cp_verify` using existing standard calibrations (this is the current desired default as specified in DMTN-222, and it's called "external verification") for a given instrument. If the tests in `cp_verify` do not pass, the Watcher should tell the observer.  I'm not quite sure how to answer your questions, though.

Reply:
In addition, I'm now also using `run_in_executor` for the butler creation and calls.